Reproducible Failure in org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests.testReadAndWriteSnapshotsThroughIndexFile

We have a failure that reproduces for me locally in `org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests.testReadAndWriteSnapshotsThroughIndexFile`:

the reproduce line is:
```
REPRODUCE WITH: ./gradlew 'null' --tests "org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests.testReadAndWriteSnapshotsThroughIndexFile" -Dtests.seed=7091BF6E1921A23F -Dtests.locale=es-CU -Dtests.timezone=Asia/Dubai -Druntime.java=14
```

Fails as:

```
[2021-02-03T15:13:09,497][INFO ][o.e.r.b.BlobStoreRepositoryTests] [testReadAndWriteSnapshotsThroughIndexFile] before test
[2021-02-03T15:13:09,497][INFO ][o.e.r.RepositoriesService] [node_s_0] update repository [test-repo]
[2021-02-03T15:13:09,507][INFO ][o.e.r.b.BlobStoreRepositoryTests] [testReadAndWriteSnapshotsThroughIndexFile] after test
REPRODUCE WITH: ./gradlew 'null' --tests "org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests.testReadAndWriteSnapshotsThroughIndexFile" -Dtests.seed=7091BF6E1921A23F -Dtests.locale=es-CU -Dtests.timezone=Asia/Dubai -Druntime.java=14

org.elasticsearch.repositories.RepositoryException: [test-repo] Failed to execute cluster state update [set safe repository generation [test-repo][0]]

	at __randomizedtesting.SeedInfo.seed([7091BF6E1921A23F:98BAA7CED14B9C73]:0)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository$8.onFailure(BlobStoreRepository.java:1870)
	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.onFailure(MasterService.java:497)
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.notifyFailedTasks(MasterService.java:430)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:204)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:140)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:669)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalStateException: Tried to update repo generation to [0] but saw unexpected generation in state [RepositoryMetadata{test-repo}{Kz4XrW6wTk29rKkY9sjYGA}{fsLike}{{"client.type":"node","cluster.initial_master_nodes":["node_s_0"],"cluster.name":"single-node-cluster-TEST_WORKER_VM=[--not-gradle--]-CLUSTER_SEED=[6282280897209334108]-HASH=[5A919654076]","cluster.routing.allocation.disk.watermark.flood_stage":"1b","cluster.routing.allocation.disk.watermark.high":"1b","cluster.routing.allocation.disk.watermark.low":"1b","discovery.seed_hosts":[],"indices.breaker.total.use_real_memory":"false","location":"/tmp/org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests_7091BF6E1921A23F-001/tempDir-002/repo/NrrLykXSUu","node.id.seed":"4772828815325206090","node.name":"node_s_0","node.processors":"1","node.roles":["data","remote_cluster_client","ingest","master"],"path.home":"/tmp/org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests_7091BF6E1921A23F-001/tempDir-002","path.logs":"/tmp/org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests_7091BF6E1921A23F-001/tempDir-002/logs","path.repo":["/tmp/org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests_7091BF6E1921A23F-001/tempDir-002/repo"],"path.shared_data":"/tmp/org.elasticsearch.repositories.blobstore.BlobStoreRepositoryTests_7091BF6E1921A23F-001","transport.port":"13301-13400","transport.type":"mock-nio"}}{0}{0}]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository$8.execute(BlobStoreRepository.java:1849)
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:686)
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:308)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:203)
	... 9 more
```

This only fails if you run the full test suit, not if you loop the individual test. It's a result of the interaction between tests because we're reusing the repository across tests and manually manipulating it at the same time. I'll fix the test.

--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -1414,7 +1414,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                 };
                 threadPool.generic().execute(() -> doGetRepositoryData(
                         ActionListener.wrap(repoData -> clusterService.submitStateUpdateTask(
-                                "set safe repository generation [" + metadata.name() + "][" + repoData + "]",
+                                "set initial safe repository generation [" + metadata.name() + "][" + repoData.getGenId() + "]",
                                 new ClusterStateUpdateTask() {
                                     @Override
                                     public ClusterState execute(ClusterState currentState) {
@@ -1441,6 +1441,8 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
 
                                     @Override
                                     public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                                        logger.trace("[{}] initialized repository generation in cluster state to [{}]",
+                                                metadata.name(), repoData.getGenId());
                                         // Resolve listeners on generic pool since some callbacks for repository data do additional IO
                                         threadPool.generic().execute(() -> {
                                             final ActionListener<RepositoryData> existingListener;
@@ -1449,6 +1451,8 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                                                 repoDataInitialized = null;
                                             }
                                             existingListener.onResponse(repoData);
+                                            logger.trace("[{}] called listeners after initializing repository to generation [{}]"
+                                                    , metadata.name(), repoData.getGenId());
                                         });
                                     }
                                 }), onFailure)));
@@ -1719,6 +1723,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
      */
     protected void writeIndexGen(RepositoryData repositoryData, long expectedGen, Version version,
                                  Function<ClusterState, ClusterState> stateFilter, ActionListener<RepositoryData> listener) {
+        logger.trace("[{}] writing repository data on top of expected generation [{}]", metadata.name(), expectedGen);
         assert isReadOnly() == false; // can not write to a read only repository
         final long currentGen = repositoryData.getGenId();
         if (currentGen != expectedGen) {
@@ -1776,6 +1781,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
 
                 @Override
                 public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    logger.trace("[{}] successfully set pending repository generation to [{}]", metadata.name(), newGen);
                     setPendingStep.onResponse(newGen);
                 }
             });
@@ -1872,6 +1878,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
 
                     @Override
                     public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                        logger.trace("[{}] successfully set safe repository generation to [{}]", metadata.name(), newGen);
                         cacheRepositoryData(repoDataToCache, newGen);
                         threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.supply(listener, () -> {
                             // Delete all now outdated index files up to 1000 blobs back from the new generation.

--- a/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java
+++ b/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java
@@ -183,14 +183,13 @@ public class BlobStoreRepositoryTests extends ESSingleNodeTestCase {
         assertThat(repository.readSnapshotIndexLatestBlob(), equalTo(expectedGeneration + 2L));
     }
 
-    public void testRepositoryDataConcurrentModificationNotAllowed() {
+    public void testRepositoryDataConcurrentModificationNotAllowed() throws Exception {
         final BlobStoreRepository repository = setupRepo();
 
         // write to index generational file
         RepositoryData repositoryData = generateRandomRepoData();
         final long startingGeneration = repositoryData.getGenId();
-        final PlainActionFuture<RepositoryData> future = PlainActionFuture.newFuture();
-        repository.writeIndexGen(repositoryData, startingGeneration, Version.CURRENT, Function.identity(), future);
+        writeIndexGen(repository, repositoryData, startingGeneration);
 
         // write repo data again to index generational file, errors because we already wrote to the
         // N+1 generation from which this repository data instance was created

